pkg/relayer/session/persistence.go

```diff
@@ -0,0 +1,261 @@
+
+		claimWindowOpenHeight := sharedtypes.GetClaimWindowOpenHeight(sharedParams, sessionEndHeight)
+		claimWindowCloseHeight := sharedtypes.GetClaimWindowCloseHeight(sharedParams, sessionEndHeight)
+		proofWindowCloseHeight := sharedtypes.GetProofWindowCloseHeight(sharedParams, sessionEndHeight)
+
+		// Create a contextual logger for this specific session
+		sessionLogger := logger.With(
+			"session_id", sessionId,
+			"supplier_operator_address", supplierOperatorAddress,
+			"session_end_height", sessionEndHeight,
+			"claim_window_close", claimWindowCloseHeight,
+			"proof_window_close", proofWindowCloseHeight,
+			"current_height", height,
+		)
+
+		var claim *prooftypes.Claim
+		if height >= claimWindowOpenHeight {
+			foundClaim, err := rs.proofQueryClient.GetClaim(ctx, supplierOperatorAddress, sessionId)
+			if err != nil && status.Convert(err).Code() != codes.NotFound {
+				sessionLogger.Error().Err(err).Msgf("failed to query claim for session %s", sessionId)
+				continue
+			} else {
+				claim = foundClaim.(*prooftypes.Claim)
+			}
+		}
+
+		// Check if session is expired (past proof submission window)
+		// These sessions can no longer reach settlement and should be deleted
+		if height > proofWindowCloseHeight {
+			// A claim was created onchain, but the session is outdated.
+			if claim != nil {
+				sessionLogger.Warn().
+					Msg("Session is outdated, but a claim exists onchain, expecting the supplier to be slashed")
+			}
+
+			if storeErr := rs.deletePersistedSessionTree(persistedSMT); storeErr != nil {
+				sessionLogger.Error().Err(storeErr).Msg("failed to delete outdated session tree, skipping")
+				continue
+			}
+
+			sessionLogger.Info().Msg("Session deleted: expired past proof submission window")
+
+			continue
+		}
+
+		// For sessions past the claim window, check if a claim exists onchain
+		// If no claim exists, the session can't progress to settlement and should be deleted
+		if height > claimWindowCloseHeight {
+			// No claim was created onchain, delete the session from the store.
+			if claim == nil {
+				if storeErr := rs.deletePersistedSessionTree(persistedSMT); storeErr != nil {
+					sessionLogger.Error().Err(storeErr).Msg("failed to delete outdated session without claim, skipping")
+					continue
+				}
+
+				sessionLogger.Info().Msg("deleting session has no claim created onchain, deleting session metadata")
+				continue
+			}
+		}
+
+		// The session has still a chance to reach settlement.
+		sessionTree, treeErr := ImportSessionTree(persistedSMT, claim, rs.storesDirectory, rs.logger)
+		if treeErr != nil {
+			sessionLogger.Error().Err(treeErr).Msg("failed to import session tree")
+			continue
+		}
+
+		// Insert the session tree into the in-memory map to be processed by the observable pipeline.
+		if ok := rs.insertSessionTree(sessionTree); !ok {
+			sessionLogger.Error().Msg("the session tree already exists, skipping")
+		}
+	}
+
+	rs.proveClaimedSessions(ctx)
```

Comment:
Can you #PUC why we're doing this here and why the normal proof pipeline is not enough?

Reply:
Done. I also added this to the claim steps of the pipeline
```
		// TODO_POST_MAINNET(red-0ne): Support skipping sessionTrees that have already their
		// claim created and are ready to be proven.
```